bookchain-miner/src/queue.hpp

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <queue>
+
+namespace bookchain {
+
+template <typename T>
+class ThreadsafeQueue {
+public:
+    void push(T& element);
+    T pop();
+
+    bool empty();
+
+private:
+    std::queue<T> _queue;
+    std::mutex _queueMutex;
+};
+
+template <typename T>
+using sharedTSQueue = std::shared_ptr<ThreadsafeQueue<T>>;
+
+template <typename T>
+void ThreadsafeQueue<T>::push(T& element) {
+    std::lock_guard<std::mutex> guard(this->_queueMutex);
```

Comment:
Til you don't need to manually release mutexes in c++ :+1:

This language is so different in industry vs academics lol

Reply:
I think it's cause in academics they typically use C or like C style C++. Once you start to actual use the modern C++ features that separate it from C things become much nicer.